server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -242,59 +262,93 @@ public Object persist(final Committer committer) throws InterruptedException
-   * transaction failure with commit metadata.
+   * @param segmentsAndMetadata
+   * @return
+   * @throws InterruptedException
    */
-  public SegmentsAndMetadata finish(
-      final TransactionalSegmentPublisher publisher,
-      final Committer committer
-  ) throws InterruptedException
+  public ListenableFuture<SegmentsAndMetadata> registerHandoff(SegmentsAndMetadata segmentsAndMetadata)
+      throws InterruptedException
   {
-    final SegmentsAndMetadata segmentsAndMetadata = publishAll(publisher, wrapCommitter(committer));
-
-    if (segmentsAndMetadata != null) {
-      final long giveUpAt = handoffConditionTimeout > 0
-                            ? System.currentTimeMillis() + handoffConditionTimeout
-                            : 0;
+    if (segmentsAndMetadata == null) {
+      return Futures.immediateFuture(null);
 
-      log.info("Awaiting handoff of segments: [%s]", Joiner.on(", ").join(appenderator.getSegments()));
+    } else {
+      final Object waitForHandoffMonitorStart = new Object();
 
-      synchronized (handoffMonitor) {
-        while (!appenderator.getSegments().isEmpty()) {
+      final ListenableFuture<SegmentsAndMetadata> future = handoffExecutor.submit(
+          () -> {
+            final List<SegmentIdentifier> waitingSegmentIdList = segmentsAndMetadata.getSegments().stream()
+                                                                                    .map(SegmentIdentifier::fromDataSegment)
+                                                                                    .collect(Collectors.toList());
+            log.info("Awaiting handoff of segments: [%s]", waitingSegmentIdList);
 
-          if (giveUpAt == 0) {
-            handoffMonitor.wait();
-          } else {
-            final long remaining = giveUpAt - System.currentTimeMillis();
-            if (remaining > 0) {
-              handoffMonitor.wait(remaining);
-            } else {
-              throw new ISE(
-                  "Segment handoff wait timeout. Segments not yet handed off: [%s]",
-                  Joiner.on(", ").join(appenderator.getSegments())
-              );
+            synchronized (waitForHandoffMonitorStart) {
+              waitForHandoffMonitorStart.notify();
+            }
+            synchronized (handoffMonitor) {
+              while (waitingSegmentIdList.stream().anyMatch(appenderator::containsSegment)) {
+                handoffMonitor.wait();
+              }
             }
+
+            log.info("All segments handed off.");
+
+            return new SegmentsAndMetadata(
+                segmentsAndMetadata.getSegments(),
+                ((FiniteAppenderatorDriverMetadata) segmentsAndMetadata.getCommitMetadata()).getCallerMetadata()
+            );
           }
-        }
+      );
+
+      // wait for handoffMonitor starts
+      log.info("Waiting handoff monitor starts");
+      synchronized (waitForHandoffMonitorStart) {
+        waitForHandoffMonitorStart.wait(handoffConditionTimeout);
```

Comment:
`Object.wait` should be called in a loop -- see https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#wait--